Refactor `FrameRefType` and add `Clear` reference type

bjoeris opened the pull request:
This change simplifies the `FrameRefType` enum by removing the separate states and
transitions, and replacing them with a set of reference types, together
with functions that compose those reference types.

This change also introduces a `Clear` reference type, which is used to represent
an access to the resource that completely overwrites the previous contents; this
is in contrast with a `Write` access that may leave some of the previous
contents untouched.

This change has been tested on Vulkan, but not with the other APIs. I am a little nervous that this may break something in another API.

Commit:
Refactor `FrameRefType` and add `Clear` reference type

This change simplifies the `FrameRefType` enum by removing the separate states and
transitions, and replacing them with a smaller set of reference types, together
with functions that compose those reference types.

This change also introduces a `Clear` reference type, which is used to represent
an access to the resource that completely overwrites the previous contents; this
is in contrast with a `Write` access that may leave some of the previous
contents untouched.

baldurk commented on renderdoc/core/resource_manager.h:
```diff
@@ -40,18 +40,22 @@ using std::map;
-  // Inputs
-  eFrameRef_Read,
-  eFrameRef_Write,
+// Compose frame refs that occur in a known order.
```
It would be helpful to ASCII-art or comment in prose here the state machine you're constructing, as reference.

baldurk commented on renderdoc/core/resource_manager.cpp:
```diff
@@ -51,44 +51,59 @@ void SetReplayResourceIDs()
+  if(first == eFrameRef_ReadBeforeWrite || second == eFrameRef_ReadBeforeWrite)
+    // Conservatively assume the ReadBeforeWrite occurs before the other ref
+    return eFrameRef_ReadBeforeWrite;
+  if((first == eFrameRef_Read) ^ (second == eFrameRef_Read))
```
I think `if((first == eFrameRef_Read) != (second == eFrameRef_Read))` should generate identical assembly and IMO is a bit clearer about the intention here.

baldurk commented on renderdoc/core/resource_manager.cpp:
```diff
@@ -51,44 +51,59 @@ void SetReplayResourceIDs()
+  if((first == eFrameRef_Clear) || (second == eFrameRef_Clear))
+    // One ref is a Clear, and the other is a Clear or Write.
+    // Whichever order these occur in will finish in Clear state
+    return eFrameRef_Clear;
```
I think I've convinced myself that this case is fine but it could maybe do with a bit of rephrasing on the comment. If I follow your intentions correctly then "[it] will finish in Clear state" is a bit misleading because we're composing these to find the _starting_ state. i.e. a `ReadBeforeWrite` followed by a `Clear` sounds like it would 'finish' in a `Clear`, but we're thinking in the opposite direction to find out its first use in the frame, so we'd consider the composition to be 'it was first used as `ReadBeforeWrite`'. Instead we're trying to indicate that a partial write followed by a complete write, without any intervening or proceeding reads, is as good as a complete write.

This might also be covered by a comment in the header explaining exactly what these states are going to mean going forward.

baldurk commented on renderdoc/core/resource_manager.cpp:
```diff
@@ -51,44 +51,59 @@ void SetReplayResourceIDs()
+    return true;
+  }
+  else
+    refit->second = ComposeFrameRefs(refit->second, refType);
```
Prefer `{}` braces around all cases in an if/else if they're present around any one.

baldurk reviewed (changes requested):
This looks good, I've made some comments mostly only wanting further clarification on exactly what each state means and how they're composed. I fully admit my fault that this wasn't documented well before but since we're changing it and since it will be a bit more involved/impactful now it'd be good to clarify what everything means. You've also got a couple of CI issues to sort out.

In terms of testing other APIs, I can take care of that. I don't think this should be a problem though - the only bit of code that used to be using this was the D3D11 code that would skip initial contents if they were large and it wasn't in `eFrameRef_ReadBeforeWrite` state. Since I removed that code in v1.2 I don't think anything cares as long as it can tell read-only access from write access, which you're preserving just fine.

bjoeris commented:
I think I have addressed all comments, and attempted to fix the CI issue.

baldurk commented on renderdoc/core/resource_manager.h:
```diff
@@ -36,23 +36,80 @@
+  eFrameRef_Clear = 2,
+
+  // Read from the resource;
+  // As a state, this represents a read that colud have seen the resource's
```
typo 'colud'

baldurk commented on renderdoc/core/resource_manager.h:
```diff
@@ -36,23 +36,80 @@
+  // Read followed by a write;
+  // As a state, this represents a read that could have seen the resource
+  // initial contents, followed by a write that could have modified that
+  // initial contents; therefore, the initila contents will need to be reset
```
typo 'initila'

baldurk commented on renderdoc/core/resource_manager.h:
```diff
@@ -36,23 +36,80 @@
+  // initial contents might still be visible to later reads.
+  eFrameRef_Write = 1,
+
+  // Clear the entire resuorce.
```
typo 'resuorce'

baldurk commented on renderdoc/core/resource_manager.h:
```diff
@@ -36,23 +36,80 @@
+  // Initial state, no reads or writes
+  eFrameRef_None = 0,
+
+  // Write to some unkown subset of resource.
```
typo 'unkown'

baldurk reviewed (approved):
I highlighted a couple of typos but otherwise these comments are perfect, thank you very much. Approved now pending a final CI fix.

bjoeris commented:
I have fixed the typos, attempted to address the CI error, and squashed down the intermediate commit. Assuming CI passes this time I think it's good to go.

baldurk commented:
Looks like there's still a clang-format issue identified on travis.
